File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.cpp
class Solution {
public:
    int f(vector<vector<int>>& dp , string& s , string& t ,  int i , int j){
        int n = s.size();
        if(i == n || j == n) return 0;
        if(dp[i][j] != -1) return dp[i][j];
        int take = 0;
        if(s[i] == t[j]) take = 1 + f(dp , s , t  , i + 1 , j + 1);
        int notTake = max(f(dp , s , t , i + 1 , j) , f(dp , s , t , i , j + 1));
        return dp[i][j] = max(take , notTake);
    }
    int longestPalindromeSubseq(string s) {
        int n = s.size();
        vector<vector<int>> dp(n + 1, vector<int>(n + 1, -1));
        string t = s;
        reverse(begin(t) , end(t));
        return f(dp , s , t , 0 , 0);       
    }
};

File: 0516-longest-palindromic-subsequence/0516-longest-palindromic-subsequence.java
class Solution {
    public int longestPalindromeSubseq(String s) {
        int n = s.length();
        int[] cur = new int[n + 1];
        int[] next = new int[n + 1];
        for(int i = n - 1 ; i >= 0 ; i--){
            cur[i] = 1;
            for(int j = i + 1 ; j < n ; j++){
                if(s.charAt(i) == s.charAt(j)){
                    cur[j] = 2 + next[j - 1];
                }else{
                    cur[j] = Math.max(cur[j - 1] , next[j]);
                }
            }
            next = cur.clone();
        }
        return cur[n - 1];
    }
}